src/bms/lex/token.rs

```diff
@@ -686,24 +898,303 @@ impl<'a> Token<'a> {
                         message,
                     }
                 }
-                message
-                    if message.starts_with('#')
-                        && message.chars().nth(6) == Some(':')
-                        && 8 <= message.len() =>
-                {
-                    let track = command[1..4]
+                #[cfg(feature = "minor-command")]
+                charfile if charfile.starts_with("#CHARFILE") => {
+                    let path = c
+                        .next_token()
+                        .map(Path::new)
+                        .ok_or_else(|| c.make_err_expected_token("charfile filename"))?;
+                    Self::CharFile(path)
+                }
+                song if song.starts_with("#SONG") => {
+                    let id = song.trim_start_matches("#SONG");
+                    let content = c.next_line_remaining();
+                    Self::Text(ObjId::try_load(id, c)?, content)
+                }
+                exbpm if exbpm.starts_with("#EXBPM") => {
+                    let id = exbpm.trim_start_matches("#EXBPM");
+                    let v = c
+                        .next_token()
+                        .ok_or_else(|| c.make_err_expected_token("exbpm value"))?;
+                    let v = Decimal::from_fraction(
+                        GenericFraction::from_str(v)
+                            .map_err(|_| c.make_err_expected_token("f64"))?,
+                    );
+                    Self::BpmChange(ObjId::try_load(id, c)?, v)
+                }
+                #[cfg(feature = "minor-command")]
+                "#BASEBPM" => {
+                    let v = c
+                        .next_token()
+                        .ok_or_else(|| c.make_err_expected_token("basebpm value"))?;
+                    let v = Decimal::from_fraction(
+                        GenericFraction::<BigUint>::from_str(v)
+                            .map_err(|_| c.make_err_expected_token("f64"))?,
+                    );
+                    Self::BaseBpm(v)
+                }
+                #[cfg(feature = "minor-command")]
+                stp if stp.starts_with("#STP") => {
+                    let line = c.next_line_remaining();
+                    let part = line.trim();
+                    if part.is_empty() {
+                        return Err(c.make_err_expected_token("stp definition"));
+                    }
+                    // Parse xxx.yyy zzzz
+                    let (xy, ms) = if let Some((xy, ms)) = part.split_once(' ') {
+                        (xy, ms)
+                    } else if let Some((xy, ms)) = part.split_once('\t') {
+                        (xy, ms)
+                    } else if let Some((xy, ms)) = part.split_once('\u{3000}') {
+                        (xy, ms)
+                    } else {
+                        return Err(c.make_err_expected_token("stp format xxx.yyy zzzz"));
+                    };
```

Comment:
If you want to extract  spans of `xxx.yyy` and `zzz`, you can use `c.next_token()` instead of using `next_line_remaining` then splitting with whitespaces.